### ISSUE TITLE
OptimSharded multi-device strategy causes device mismatch during validation

### ISSUE THREAD

**Issue by Dreaming-Codes** [tracel-ai/burn#4472]
## Description

When using `TrainingStrategy::MultiDevice` with `MultiDeviceOptim::OptimSharded`, validation fails with a tensor device mismatch error after the first training epoch.

## Cause

In `MultiDeviceLearningStrategy::fit()` (`burn-train/src/learner/supervised/strategies/multi/strategy.rs`), `learner.fork(main_device)` is called once at the start. During training, `OptimSharded`'s `step_multi` distributes model parameters across devices — each parameter ends up on whichever device was selected for its gradient accumulation (via `(id.val() + i) % num_devices` in `MultiGradientsParams::select()`).

After the training epoch completes, validation runs immediately via `SingleDeviceValidEpoch` **without forking the model back to `main_device`**. Since parameters are now scattered across multiple CUDA devices, any operation involving parameters on different devices (e.g., `Matmul` with LHS on `Cuda(1)` and RHS on `Cuda(2)`) panics.

## Error

```
=== Tensor Operation Error ===
  Operation: 'Matmul'
  Reason:
    1. The provided tensors are not on the same device. Lhs tensor device Cuda(1), Rhs tensor device Cuda(2).
```

Stack trace goes through: `SingleDeviceValidEpoch::run` → `InferenceStep::step` → model forward pass → `Linear::forward` → `matmul`

## Proposed Fix

I'm not sure if this is the right fix as I don't fully understand the Burn architecture and there's no documentation about distributed computing, but adding `learner.fork(main_device)` before `epoch_valid.run()` in the training loop worked for me:

```rust
// After OptimSharded training, model parameters are scattered across
// devices. Fork back to main_device before validation.
learner.fork(main_device);
```

This reassembles the model on a single device before validation. It should be safe because the next training epoch's `MultiDevicesTrainStep` will fork per-worker anyway.

## Environment

- Burn 0.20.1
- 4x NVIDIA GPUs (CUDA)
- `TrainingStrategy::MultiDevice(devices, MultiDeviceOptim::OptimSharded)`

**Comment by Dreaming-Codes** [comment]
Let me know if I should open a PR with the described fix

**Comment by laggui** [comment]
> Let me know if I should open a PR with the described fix

Yeah I think that would be the correct fix for single-device validation. Could check for `matches!(self.optim, MultiDeviceOptim::OptimSharded)` to perform the fork.
